File: jclang/src/test/java/jclang/IndexTest.java


package jclang;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.junit.jupiter.api.Test;

import java.io.File;
import java.io.PrintWriter;
import java.io.StringWriter;

import static jclang.TestUtils.createOrCompare;
import static org.junit.jupiter.api.Assertions.*;

public class IndexTest extends ClangTest {
    @NotNull
    private String getDir() {
        return getTestDataDir() + "/index/";
    }

    @NotNull
    private String getTestDeclarationsFile() {
        return getDir() + "declarations.h";
    }

    @NotNull
    private TranslationUnit indexTestDeclarations(@NotNull IndexerCallback callback) {
        Index index = Clang.INSTANCE.createIndex(false, false);
        return index.indexSourceFile(callback, getTestDeclarationsFile(), new String[]{});
    }

    @Test
    public void testIndexSourceFileWithEmptyCallback() {
        TranslationUnit unit = indexTestDeclarations(IndexerCallback.DO_NOTHING);
        assertTrue(unit.getDiagnostics().isEmpty());
    }

    @Test
    public void testStartedTranslationUnit() {
        final boolean[] started = new boolean[] {false};
        indexTestDeclarations(new AbstractIndexerCallback() {
            @Override
            public void startedTranslationUnit() {
                assertFalse(started[0]);
                started[0] = true;
            }
        });
        assertTrue(started[0]);
    }

    @Test
    public void testEnteredMainFile() {
        final File[] handle = new File[] {null};
        indexTestDeclarations(new AbstractIndexerCallback() {
            @Override
            public void enteredMainFile(@NotNull File mainFile) {
                handle[0] = mainFile;
            }
        });
        assertNotNull(handle[0]);
        assertEquals(new File(getTestDeclarationsFile()).getAbsoluteFile(), handle[0].getAbsoluteFile());
    }

    @Test
    public void testIndexDeclaration() {
        StringWriter sw = new StringWriter();
        final PrintWriter out = new PrintWriter(sw);
        indexTestDeclarations(new AbstractIndexerCallback() {
            @Override
            public void indexDeclaration(@NotNull DeclarationInfo info) {
                Cursor cursor = info.getCursor();
                IndexLocation location = info.getLocation();
                out.print(location.getLine() + ":" + location.getColumn());
                out.print(" " + cursor.getKind().getSpelling());
                out.print(" " + cursor.getType().getKind());
                out.print(" " + nonEmptyCursorSpelling(cursor));
                out.println();

                EntityInfo entityInfo = info.getEntityInfo();
                out.println("  " + entityInfo.getUSR());
                out.println("  " + entityInfo.getKind());
                String spelling = cursor.getSpelling();
                assertEquals(spelling.isEmpty() ? null : spelling, entityInfo.getName());

                if (info.isRedeclaration()) out.println("  redecl");
                if (info.isDefinition()) out.println("  def");
                if (info.isContainer()) out.println("  cnt");
                if (info.isImplicit()) out.println("  implicit");
            }
        });
        out.close();
        createOrCompare(sw.toString(), getDir() + "indexDeclaration.txt");
    }

    @NotNull
    private String nonEmptyCursorSpelling(@NotNull Cursor cursor) {
        String spelling = cursor.getSpelling();
        return spelling.isEmpty() ? "<no-name>" : spelling;
    }

    @Test
    public void testIndexObjCAttributes() {
        StringWriter sw = new StringWriter();
        final PrintWriter out = new PrintWriter(sw);
        Index index = Clang.INSTANCE.createIndex(false, false);
        index.indexSourceFile(new AbstractIndexerCallback() {
            @Override
            public void indexDeclaration(@NotNull DeclarationInfo info) {
                out.println(info.getCursor().getSpelling());
                for (IndexAttribute attribute : info.getAttributes()) {
                    IndexLocation location = attribute.getLocation();
                    out.print("  " + location.getLine() + ":" + location.getColumn());
                    out.print(" " + attribute.getKind());
                    out.print(" " + attribute.getCursor().getKind());
                    out.println();
                }
            }
        }, getDir() + "objcAttributes.h", new String[]{"-ObjC"});
        out.close();
        createOrCompare(sw.toString(), getDir() + "objcAttributes.txt");
    }

    @Test
    public void testIndexContainerInfo() {
        StringWriter sw = new StringWriter();
        final PrintWriter out = new PrintWriter(sw);
        Index index = Clang.INSTANCE.createIndex(false, false);
        index.indexSourceFile(new AbstractIndexerCallback() {
            @Override
            public void indexDeclaration(@NotNull DeclarationInfo info) {
                out.println(nonEmptyCursorSpelling(info.getCursor()));
                printContainerCursor(info.getSemanticContainer());
                printContainerCursor(info.getLexicalContainer());
                printContainerCursor(info.getDeclAsContainer());
            }

            private void printContainerCursor(@Nullable ContainerInfo container) {
                if (container == null) return;
                Cursor cursor = container.getCursor();
                out.println("  " + cursor.getKind() + " " + nonEmptyCursorSpelling(cursor));
            }
        }, getDir() + "containerInfo.h", new String[]{"-c", "-x", "c++"});
        out.close();
        createOrCompare(sw.toString(), getDir() + "containerInfo.txt");
    }
}

File: toolkit/src/test/kotlin/klang/IndexTest.kt
package klang

import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import klang.jvm.*
import java.io.File
import java.io.PrintWriter
import java.io.StringWriter

class IndexTest : StringSpec({

    afterTest { (_, _) ->
        NativePool.disposeAll()
    }

    "test IndexSourceFileWithEmptyCallback" {
        val unit = indexTestDeclarations(IndexerCallback.DO_NOTHING)
        unit.diagnostics.isEmpty() shouldBe true
    }

    "test StartedTranslationUnit" {
        var started = false
        indexTestDeclarations(object : AbstractIndexerCallback() {
            override fun startedTranslationUnit() {
                started = true
            }
        })
        started shouldBe true
    }

    "test EnteredMainFile" {
        var handle:File? = null
        indexTestDeclarations(object : AbstractIndexerCallback() {
            override fun enteredMainFile(mainFile: File) {
                handle = mainFile
            }
        })

        handle shouldNotBe null
        handle!!.absoluteFile shouldBe File(getTestDeclarationsFile()).absoluteFile

    }

    "test IndexDeclaration" {
        val sw = StringWriter()
        val out = PrintWriter(sw)
        indexTestDeclarations(object : AbstractIndexerCallback() {
            override fun indexDeclaration(info: DeclarationInfo) {
                val cursor = info.cursor
                val location = info.location
                out.print(location.getLine().toString() + ":" + location.getColumn())
                out.print(" " + cursor.kind.spelling)
                out.print(" " + cursor.type.kind)
                out.print(" " + nonEmptyCursorSpelling(cursor))
                out.println()
                val entityInfo = info.entityInfo
                out.println("  " + entityInfo.USR)
                out.println("  " + entityInfo.kind)
                val spelling = cursor.spelling
                entityInfo.name shouldBe spelling.ifEmpty { null }
                if (info.isRedeclaration) out.println("  redecl")
                if (info.isDefinition) out.println("  def")
                if (info.isContainer) out.println("  cnt")
                if (info.isImplicit) out.println("  implicit")
            }
        })
        out.close()
        createOrCompare(sw.toString(), getDir() + "indexDeclaration.txt")
    }

    "test IndexObjCAttributes" {
        val sw = StringWriter()
        val out = PrintWriter(sw)
        val index = createIndex(excludeDeclarationsFromPCH = false, displayDiagnostics = false)
        index.indexSourceFile(object : AbstractIndexerCallback() {
            override fun indexDeclaration(info: DeclarationInfo) {
                out.println(info.cursor.spelling)
                for (attribute in info.attributes) {
                    val location = attribute.location
                    out.print("  " + location.getLine() + ":" + location.getColumn())
                    out.print(" " + attribute.kind)
                    out.print(" " + attribute.cursor.kind)
                    out.println()
                }
            }
        }, getDir() + "objcAttributes.h", arrayOf("-ObjC"))
        out.close()
        createOrCompare(sw.toString(), getDir() + "objcAttributes.txt")
    }

    "test IndexContainerInfo" {
        createIndex(excludeDeclarationsFromPCH = false, displayDiagnostics = false).use { index ->
            val sw = StringWriter()
            val out = PrintWriter(sw)
            index.indexSourceFile(object : AbstractIndexerCallback() {
                override fun indexDeclaration(info: DeclarationInfo) {
                    out.println(nonEmptyCursorSpelling(info.cursor))
                    printContainerCursor(info.semanticContainer)
                    printContainerCursor(info.lexicalContainer)
                    printContainerCursor(info.declAsContainer)
                }

                private fun printContainerCursor(container: ContainerInfo?) {
                    if (container == null) return
                    val cursor = container.cursor
                    out.println("  " + cursor.kind + " " + nonEmptyCursorSpelling(cursor))
                }
            }, getDir() + "containerInfo.h", arrayOf("-c", "-x", "c++"))
            out.close()
            createOrCompare(sw.toString(), getDir() + "containerInfo.txt")
        }
    }

})

private fun nonEmptyCursorSpelling(cursor: Cursor): String {
    val spelling = cursor.spelling
    return if (spelling.isEmpty()) "<no-name>" else spelling
}

private fun getDir(): String {
    return getTestDataDir().toString() + "/index/"
}

private fun getTestDeclarationsFile(): String {
    return getDir() + "declarations.h"
}

private fun indexTestDeclarations(callback: IndexerCallback): TranslationUnit {
    createIndex(excludeDeclarationsFromPCH = false, displayDiagnostics = true).use { index ->
        return index.indexSourceFile(callback, getTestDeclarationsFile(), arrayOf())
    }
}